File: packages/ui-compose/ui/src/test/kotlin/com/kinetixui/ui/blocks/SignInBlockTest.kt
package com.kinetixui.ui.blocks

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.test.junit4.v2.createComposeRule
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import com.kinetixui.ui.KinetixButton
import com.kinetixui.ui.KinetixButtonVariant
import com.kinetixui.ui.KinetixCard
import com.kinetixui.ui.KinetixCardContent
import com.kinetixui.ui.KinetixCardDescription
import com.kinetixui.ui.KinetixCardFooter
import com.kinetixui.ui.KinetixCardHeader
import com.kinetixui.ui.KinetixCardTitle
import com.kinetixui.ui.KinetixCheckbox
import com.kinetixui.ui.KinetixInput
import com.kinetixui.ui.KinetixLabel
import com.kinetixui.ui.KinetixTheme
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

// The "Sign in" block — see TestimonialBlockTest.kt for how block fixtures work.
//
// kx-block:start
@Composable
fun SignInBlock(onSignIn: () -> Unit = {}, onGithub: () -> Unit = {}) {
    var email by remember { mutableStateOf("") }
    var password by remember { mutableStateOf("") }
    var rememberMe by remember { mutableStateOf(true) }

    KinetixCard(Modifier.width(360.dp)) {
        KinetixCardHeader {
            KinetixCardTitle("Sign in")
            KinetixCardDescription("Enter your email to sign in to your account.")
        }
        KinetixCardContent {
            Column(verticalArrangement = Arrangement.spacedBy(16.dp)) {
                KinetixLabel("Email")
                KinetixInput(email, { email = it }, placeholder = "you@example.com", keyboardType = KeyboardType.Email)
                KinetixLabel("Password")
                KinetixInput(password, { password = it }, visualTransformation = PasswordVisualTransformation())
                Row(verticalAlignment = Alignment.CenterVertically) {
                    KinetixCheckbox(rememberMe, { rememberMe = it })
                    Spacer(Modifier.width(8.dp))
                    Text("Remember me")
                }
            }
        }
        KinetixCardFooter {
            Column(Modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(8.dp)) {
                KinetixButton(onSignIn, Modifier.fillMaxWidth()) { Text("Sign in") }
                KinetixButton(onGithub, Modifier.fillMaxWidth(), variant = KinetixButtonVariant.Outline) {
                    Text("Continue with GitHub")
                }
            }
        }
    }
}
// kx-block:end

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34])
class SignInBlockTest {
    @get:Rule
    val rule = createComposeRule()

    @Test
    fun renders_the_whole_form() {
        rule.setContent { KinetixTheme(darkTheme = false) { SignInBlock() } }

        rule.onNodeWithText("Enter your email to sign in to your account.").assertExists()
        rule.onNodeWithText("Email").assertExists()
        rule.onNodeWithText("Remember me").assertExists()
        rule.onNodeWithText("Continue with GitHub").assertExists()
        // "Sign in" is deliberately not asserted: it is both the card title and the submit button, so a
        // single-node matcher would be ambiguous and an index-based one would be testing layout order
    }
}

File: packages/ui-flutter/test/blocks/sign_in_block_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:kinetix_ui/kinetix_ui.dart';

// The "Sign in" block — see testimonial_block_test.dart for how block fixtures work.
//
// kx-block:start
class SignInBlock extends StatefulWidget {
  const SignInBlock({super.key, this.onSignIn, this.onGithub});

  final VoidCallback? onSignIn;
  final VoidCallback? onGithub;

  @override
  State<SignInBlock> createState() => _SignInBlockState();
}

class _SignInBlockState extends State<SignInBlock> {
  final TextEditingController _email = TextEditingController();
  final TextEditingController _password = TextEditingController();
  bool _remember = true;

  @override
  void dispose() {
    _email.dispose();
    _password.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return KinetixCard(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: <Widget>[
          const KinetixCardHeader(
            children: <Widget>[
              KinetixCardTitle('Sign in'),
              KinetixCardDescription('Enter your email to sign in to your account.'),
            ],
          ),
          KinetixCardContent(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: <Widget>[
                const KinetixLabel('Email'),
                KinetixInput(controller: _email, placeholder: 'you@example.com'),
                const SizedBox(height: 12),
                const KinetixLabel('Password'),
                KinetixInput(controller: _password, obscureText: true),
                const SizedBox(height: 12),
                Row(
                  children: <Widget>[
                    KinetixCheckbox(
                      value: _remember,
                      onChanged: (bool v) => setState(() => _remember = v),
                    ),
                    const SizedBox(width: 8),
                    const Text('Remember me'),
                  ],
                ),
              ],
            ),
          ),
          KinetixCardFooter(
            children: <Widget>[
              Expanded(
                child: Column(
                  children: <Widget>[
                    SizedBox(
                      width: double.infinity,
                      child: KinetixButton(onPressed: widget.onSignIn, child: const Text('Sign in')),
                    ),
                    const SizedBox(height: 8),
                    SizedBox(
                      width: double.infinity,
                      child: KinetixButton(
                        onPressed: widget.onGithub,
                        variant: KinetixButtonVariant.outline,
                        child: const Text('Continue with GitHub'),
                      ),
                    ),
                  ],
                ),
              ),
            ],
          ),
        ],
      ),
    );
  }
}
// kx-block:end

void main() {
  testWidgets('renders the whole form and toggles Remember me', (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: KinetixTheme(
          brightness: Brightness.light,
          child: const Scaffold(body: SingleChildScrollView(child: SignInBlock())),
        ),
      ),
    );

    expect(find.text('Enter your email to sign in to your account.'), findsOneWidget);
    expect(find.text('Email'), findsOneWidget);
    expect(find.text('Remember me'), findsOneWidget);
    expect(find.text('Continue with GitHub'), findsOneWidget);
    // 'Sign in' is both the card title and the submit button, so it is deliberately expected twice
    expect(find.text('Sign in'), findsNWidgets(2));
  });
}
